plugin/src/main/java/org/wildfly/plugins/bootablejar/maven/goals/DevWatchBootableJarMojo.java

```diff
@@ -293,6 +338,131 @@ ModelNode waitStatus(ModelControllerClient client, String status, ModelNode op)
+            steps.add(Operations.createOperation("remove", address));
+            getLog().debug("Undeploy " + name);
+            ModelNode response = client.execute(composite);
+            getLog().debug("Undeploy " + name + " done " + response.toJSONString(true));
+        }
+
+        private boolean deploy(ModelControllerClient client, Path deploymentDir) throws Exception {
+            ModelNode composite = Operations.createCompositeOperation();
+            OperationBuilder builder = new OperationBuilder(composite, true);
+            ModelNode steps = composite.get("steps");
+            ModelNode address = new ModelNode();
+            address.add("deployment", name);
+            final ModelNode op = Operations.createOperation("add", address);
+            op.get("runtime-name").set("ROOT.war");
+            ModelNode content = op.get("content").get(0);
+            content.get("empty").set(true);
+            getLog().debug("Deploy " + name);
+            steps.add(op);
+            ModelNode addContentOp = Operations.createOperation(ADD_CONTENT, address);
+            Files.walkFileTree(deploymentDir, new FileVisitor<Path>() {
+                private int index = 0;
+                private int stream = 0;
+
+                @Override
+                public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                    builder.addFileAsAttachment(file);
+                    getLog().debug("Sending file " + file + " to " + deploymentDir.relativize(file).toString() + " with index " + stream);
+                    addContentOp.get(CONTENT).get(index).get(INPUT_STREAM_INDEX).set(stream++);
+                    addContentOp.get(CONTENT).get(index).get(TARGET_PATH).set(deploymentDir.relativize(file).toString());
+                    index++;
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOException {
+                    return FileVisitResult.CONTINUE;
+                }
+
+                @Override
+                public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
+                    return FileVisitResult.CONTINUE;
+                }
+            });
+            steps.add(addContentOp);
+            steps.add(Operations.createOperation("deploy", address));
+            ModelNode reply = client.execute(builder.build());
+            getLog().debug("Deploy " + name + " done " + reply.toJSONString(true));
+            return "success".equals(reply.get("outcome").asString());
+        }
+
+        ModelNode waitStatus(ModelControllerClient client, String status, ModelNode op) throws Exception {
```

Comment:
I think that we should share the logic with the localDeploymentController, introduce an AbstractDeploymentController?